server/util/fspath/fspath.go

```diff
@@ -0,0 +1,47 @@
+package fspath
+
+import (
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+)
+
+// Whether the filesystems most commonly used on the current platform are
+// case-insensitive.
+//
+// TODO: actually check the filesystem - it's possible (but uncommon) to have
+// case-sensitive filesystems on these platforms.
+var isCaseInsensitive = runtime.GOOS == "darwin" || runtime.GOOS == "windows"
+
+// Key is a representation of a filesystem path suitable for use as a map key.
+// It properly handles case-insensitive filesystems, in which paths with
+// different cases should be considered equal.
+type Key struct {
+	value string
+}
+
+// NewKey returns a new Key for the given path.
+func NewKey(path string) Key {
+	if isCaseInsensitive {
+		return Key{value: strings.ToLower(path)}
```

Comment:
It probably doesn't matter in practice, but it's worth keeping in mind that this scheme isn't 100% sound: the Unicode case folding data in Go may differ from how macOS or Windows actually normalize paths. It would be cleaner to go through FS APIs such as `os.SameFile()`, but I'm totally fine with the current approach.

Reply:
I'm not sure if `os.SameFile` is practical given that we need to do quick string comparisons via map lookups - I guess if we hit OS-specific quirks we will need to replicate the quirks here.

Reply:
os.SameFile requires both entries to be present so that we can call `stat` on them. It's not applicable here.